File: my-folder/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    // abcbad
    // ab[cbad]
    public int lengthOfLongestSubstring(String s) {
        int maxLen = 0;
        int left = 0;
        // char -> right most index
        Map<Character, Integer> map = new HashMap<>();
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if(map.containsKey(ch)){
                left = Math.max(left, map.get(ch) + 1);
            }
            map.put(ch, i);
            maxLen = Math.max(maxLen, i-left+1);
        }
        return maxLen;
    }
}

File: my-folder/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        n = len(s)
        left, right = 0, 0
        Set = set()
        while right < n:
            entry = s[right]
            while entry in Set:
                Set.remove(s[left])
                left += 1
            Set.add(entry)
            ans = max(ans, len(Set))
            right += 1
        return ans
                
